pkg/option: add option to configure BPF lbmap size

fristonio opened the pull request:
Add a command-line flag to cilium-agent to configure BPF lbmap size
with a user-provided value. This commit adds a new flag called
`bpf-lb-map-max` for this. The default value of max entries in BPF lbmap is 65536.

Running cilium-agent with `--bpf-lb-map-max 131072` gives the maps with the following sizes.

![image](https://user-images.githubusercontent.com/21292343/89995787-9b37dc80-dca7-11ea-876c-119b1ea54707.png)

FIxes #12815

fristonio committed:
pkg/option: add option to configure BPF lbmap size

Add a command line flag to cilium-agent to configure BPF lbmap size
according to a user provided value. This commit adds a new flag called
`bpf-lb-map-max`. The default value of max entries in BPF lbmap is
65536.

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

brb commented on daemon/cmd/daemon_main.go:
```diff
@@ -815,6 +815,9 @@ func init() {
 	flags.Int(option.FragmentsMapEntriesName, defaults.FragmentsMapEntries, "Maximum number of entries in fragments tracking map")
 	option.BindEnv(option.FragmentsMapEntriesName)
 
+	flags.Int(option.LBMapEntriesName, defaults.LBMapEntries, "Maximum number of entries in Cilium BPF lbmap")
```
I'd keep the default in `pkg/lbmap` package.

fristonio replied:
I implemented it similar to how we populate `option.Config.PolicyMapEntries`. We use default value of PolicyMapEntries from `pkg/defaults`(https://github.com/cilium/cilium/blob/master/daemon/cmd/daemon_main.go#L714-L715). Should I change that as well, so that we maintain consistency?

brb replied:
> Should I change that as well, so that we maintain consistency?

If @cilium/policy is OK with it, then yes please.

brb commented on pkg/defaults/defaults.go:
```diff
@@ -397,4 +397,8 @@ const (
 	// API groups using Discovery API.
 	K8sEnableLeasesFallbackDiscovery = false
+
+	// LBMapEntries contains the default number of entries that are allowed in cilium
```
Could you extend the comment by saying that the default is also used for the LB backend and affinity maps?

brb reviewed (commented):
Thanks!

Unrelated to your PR: shall the LB map size be a subject to the dynamic BPF resizing? /cc @tklauser

fristonio committed:
install/kubernetes: add helm option to override lbmap max entries

This commits adds a new helm bpf option `global.bpf.lbMapMax` that can
be used to configure the max entries in LB service, backend and affinity
maps. Extend KubeProxy free gsg to point out the use of this helm
option when user needs to scale the cluster beyond a default preconfigured
limit.

Signed-off-by: Deepesh Pathak <deepshpathak@gmail.com>

brb commented on pkg/option/config.go:
```diff
@@ -802,6 +802,15 @@ const (
+	LBMapEntriesName = "bpf-lb-map-max"
+
+	// LBMapMinEntries defines the minimum lbmap limit.
+	LBMapMinEntries = 1 << 10 // 1Ki entries
```
Where does the min entries limit come from? I think some users might want to go below this limit (think of embedded devices with very limited memory).

fristonio replied:
I discussed this on slack with @aanm. This is similar to the limits on CT/NAT table - [LimitTable{Max,Min}](https://github.com/cilium/cilium/blob/master/pkg/option/config.go#L490-L494). We discussed that this could be a good starting point. But if it makes sense like for the case you mentioned we can decide on a more suitable limit.

brb replied:
Can we then just remove those artificial limits? I think the CT/NAT limits might be used by the dynamic scaling.

borkmann replied:
@fristonio @aanm  why not just check for invalid configs like map size of 0 and the rest is up to the user?

aanm replied:
@borkmann @brb let's address in a different PR, this was essentially introduced by @joestringer  https://github.com/cilium/cilium/commit/ec07900a49bf96b2f47b6ecbc0b4c22c7ca4de46#diff-b72a2fd7a91a97fb4b234e7919334832R1090

brb replied:
Let's just remove the lbmap size checks in this PR.

fristonio replied:
Cool, I will remove the checks on `lbmap` size from this PR and will remove the defaults for these sizes from `pkg/defaults`.

aanm commented on pkg/option/config_test.go:
```diff
@@ -434,6 +441,26 @@ func TestCheckMapSizeLimits(t *testing.T) {
+			want: sizes{
+				LBMapEntries: LBMapMaxEntries + 1,
+				WantErr:      true,
+			},
```
Are we returning an error because of the `LBMapEntries` or because of the other map sizes?

fristonio replied:
Yeah, because of LBMapEntries. Similar to what we do for `PolicyMapEntries` - https://github.com/cilium/cilium/blob/3466e7157af1949f81fd4ad78f24a4a7817c77c4/pkg/option/config_test.go#L424-L443

brb commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -706,6 +706,31 @@ and therefore not affecting any application pod ``bind(2)`` requests anymore. In
+
+.. note::
+
+   Scaling size of Cilium's BPF LB map comes with a limitation that the Reverse NAT
```
This might confuse users - for each service we create a surrogate entry which doesn't point to any backend, but it is used to determine a number of backends (for the second lookup). So, for a service w/o any backend we have 1 entry with the same rev NAT ID, while w/ N backends - N + 1.

fristonio replied:
As discussed on slack with @brb, since revNAT ID limit will be hit later than the LB map max size(unless user runs services without any backend) it might make sense to remove this note. @borkmann what are your thoughts on this?

borkmann replied:
Yes, ACK for removing this note.

brb commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -31,9 +32,9 @@ import (
-	MaxEntries = 65536
+var (
+	// MaxEntries is the maximum number of entries in each hashtable
+	MaxEntries = defaults.LBMapEntries
```
Not sure whether it makes sense to have another indirection by storing the actual number in `defaults`. I was suggesting to have it here.

fristonio replied:
Yeah, I think it makes sense. I will update the PR.

aanm reviewed (approved):
Nothing else besides what Martynas point it out.
